File: Competitive Programming 3/Section 1.3.3/IOI '09 - Plovdiv, Bulgaria - Garage/Python/main.py
from queue import *
from heapq import heapify,heappush,heappop 

def main():
    n,m = map(int,input().split())
    costPerKilo,available,weight,slotTaken=[],[],[],[0]*m
    heapify(available)
    cost=0
    waitingLine = Queue()
    for i in range(0,n):
        costPerKilo.append(int(input()))
        heappush(available,i)
    for i in range(0,m):
        weight.append(int(input()))
    for i in range(0,m<<1):
        car = int(input())
        if car < 0:
            heappush(available,slotTaken[abs(car)-1])
        else:
            waitingLine.put(car-1)
        while available and not waitingLine.empty():
            slot = heappop(available)
            carRemoved = waitingLine.get()
            cost += costPerKilo[slot] * weight[carRemoved]
            slotTaken[carRemoved] = slot
    print(cost)
main()

File: Competitive Programming 3/Section 1.3.3/IOI '09 - Plovdiv, Bulgaria - Garage/C++/main.cpp
#include <bits/stdc++.h>

#define maxCars 2002

using namespace std;

int costPerKilo[102],weight[maxCars],slotTaken[maxCars];
priority_queue< int,vector<int>,greater<int> > available;
queue<int> waitingLine;
int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    int n,m,cost=0;
    cin >> n >> m ;
    for(int i = 1 ; i <= n ; i ++){
        cin >> costPerKilo[i];
        available.push(i);
    } 
    for(int i = 1 ; i <= m ; i ++){
        cin >> weight[i];
    }
    for(int m2 = m<<1; m2;m2--){
        int car;
        cin >> car;
        if(car < 0)
            available.push(slotTaken[-car]);
        else{
            waitingLine.push(car);
        }
        while(!available.empty()&&!waitingLine.empty()){
            int slot = available.top(), carRemoved = waitingLine.front();
            available.pop(),waitingLine.pop();
            cost += costPerKilo[slot] * weight[carRemoved];
            slotTaken[carRemoved] = slot;
        }
    }
    cout << cost << "\n";
    return 0;
}